### PR TITLE
Templated _d_arrayshrinkfit implementation

### PR DESCRIPTION
**Pull request by yashjha123**
Draft PR: Possible duplicate of #21152  This PR completely templatizes the _d_arrayshrinkfit hook and adds more unittests that cover good edge cases. This PR may be deleted if there is a conflict, and is therefore a draft.

This PR introduces a template version of the _d_arrayshrinkfit DRuntime Hook, often used as a call with `assumesafeappend`. 

Changes Made
1. Templated Hook: Implement a new template version of the hook in DRuntime
2. Change the lowering in the compiler (capacity.d) to use the templated version
3. Created unit-tests to check for any issues (tests all the edge cases I can think of)
4. Removed the old hook from DRuntime

### CODE REVIEW

File: druntime/src/core/internal/array/capacity.d [hunk 1]
```diff
@@ -86,3 +158,89 @@ template _d_arraysetlengthTImpl(Tarr : T[], T)
...
+    assert(arr.length == 16);
+    foreach (int i; arr)
+        assert(i == int.init);
+
+    shared S[] arr2;
+    _d_arraysetlengthTImpl!(typeof(arr2))._d_arraysetlengthT(arr2, 16);
+    assert(arr2.length == 16);
+    foreach (s; arr2)
+        assert(s == S.init);
+}
```
Comment (teodutu) [inline druntime/src/core/internal/array/capacity.d:180]: Why did you add this test? It's about `_d_arraysetlengthT`.

File: druntime/src/core/internal/array/capacity.d [hunk 2]
```diff
@@ -8,11 +8,83 @@
...
+Shrink the "allocated" length of an array to be the exact size of the array.
+
+It doesn't matter what the current allocated length of the array is, the
+user is telling the runtime that he knows what he is doing.
+
+Params:
+    ti = `TypeInfo` of array type
+    arr = array to shrink. Its `.length` is element length, not byte length, despite `void` type
+*/
+void _d_arrayshrinkfit(T)(void[] arr) nothrow
```
Comment (teodutu) [inline druntime/src/core/internal/array/capacity.d:33]:
```suggestion
void _d_arrayshrinkfit(T)(T[] arr) @trusted
```
The compiler infers `pure`, `@safe` and `nothrow` for templates. But hooks are almost never safe, so mark them as `@trusted`.

File: druntime/src/core/internal/array/capacity.d [hunk 3]
```diff
@@ -8,11 +8,83 @@
...
+        return;
+
+    // align the array.
+    auto offset = arr.ptr - curArr.ptr;
+    auto cursize = curArr.length - offset;
+    if (cursize <= reqsize)
+        // invalid situation, or no change.
+        return;
+
+    static if (is(T == struct) && hasElaborateDestructor!T)
```
Comment (teodutu) [inline druntime/src/core/internal/array/capacity.d:54]:
```suggestion
    static if (hasElaborateDestructor!T)
```
Don't limit this to structs. Classes can have destructors too.

File: druntime/src/core/internal/array/capacity.d [hunk 4]
```diff
@@ -8,11 +8,83 @@
...
+    {
+        try
+        {
+            finalize_array!T(arr.ptr + reqsize, cursize - reqsize);
```
Comment (teodutu) [inline druntime/src/core/internal/array/capacity.d:58]: You can inline the function call here

File: druntime/src/core/internal/array/capacity.d [hunk 5]
```diff
@@ -8,11 +8,83 @@
...
+    import object: destroy;
+
+    // Due to the fact that the delete operator calls destructors
+    // for arrays from the last element to the first, we maintain
+    // compatibility here by doing the same.
+    auto tsize = T.sizeof;
+    for (auto curP = p + size - tsize; curP >= p; curP -= tsize)
+    {
+        // call destructor
+        destroy(*cast(T*)curP);
```
Comment (teodutu) [inline druntime/src/core/internal/array/capacity.d:80]:
```suggestion
        (cast(T*) curP).__xdtor();
```
This calls the destructor directly since you already know the type has one.

File: druntime/src/core/internal/array/capacity.d [hunk 6]
```diff
@@ -86,3 +158,89 @@ template _d_arraysetlengthTImpl(Tarr : T[], T)
...
+    C[] arr = new C[10];
+    foreach (ref c; arr)
+        c = new C();
+
+    arr.length = 3;
+    _d_arrayshrinkfit!C(arr);
+    assert(arr.length == 3);
+    foreach (c; arr)
+        assert(c !is null && c.x == 5);
+}
```
Comment (teodutu) [inline druntime/src/core/internal/array/capacity.d:246]: The essence of `_d_arrayshrinkfit` is that any appends made after calling it will not reallocate. Testing if the elements are the same is fine, but the way to test this fully is to check if the array's `.ptr` is the same before and after calling the hook.

File: druntime/src/core/internal/array/capacity.d [hunk 7]
```diff
@@ -8,11 +8,83 @@
...
 module core.internal.array.capacity;
+import core.exception : onFinalizeError;
 
 // HACK: `nothrow` and `pure` is faked.
 private extern (C) void[] _d_arraysetlengthT(const TypeInfo ti, size_t newlength, void[]* p) nothrow pure;
 private extern (C) void[] _d_arraysetlengthiT(const TypeInfo ti, size_t newlength, void[]* p) nothrow pure;
 
+extern(C) {
+    bool gc_shrinkArrayUsed(void[] slice, size_t existingUsed, bool atomic) nothrow;
+    void[] gc_getArrayUsed(void *ptr, bool atomic) nothrow;
```
Comment (teodutu) [inline druntime/src/core/internal/array/capacity.d:19]:
```suggestion
    bool gc_shrinkArrayUsed(void[] slice, size_t existingUsed, bool atomic) pure nothrow;
    void[] gc_getArrayUsed(void *ptr, bool atomic) pure nothrow;
```
You'll most likely have to lie about purity too.

**Review by teodutu — CHANGES_REQUESTED** [review]
This is a good start. Apply my suggestions and keep it up!
